saml/src/main/java/com/linecorp/armeria/server/saml/JwtBasedSamlRequestIdManager.java

```diff
@@ -0,0 +1,99 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.saml;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.JWTVerifier;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.google.common.base.MoreObjects;
+
+import io.netty.util.internal.ThreadLocalRandom;
+
+/**
+ * A {@link SamlRequestIdManager} implementation based on JSON Web Tokens specification.
+ * See <a href="https://jwt.io/">JSON Web Tokens</a> for more information.
+ */
+final class JwtBasedSamlRequestIdManager implements SamlRequestIdManager {
+
+    private static final String CLAIM_NAME_SALT = "salt";
+
+    private final String issuer;
+    private final int validSeconds;
+    private final int leewaySeconds;
+
+    private final Algorithm algorithm;
+    private final JWTVerifier verifier;
+
+    JwtBasedSamlRequestIdManager(String issuer, Algorithm algorithm,
+                                 int validSeconds, int leewaySeconds) {
+        this.issuer = requireNonNull(issuer, "issuer");
+        this.algorithm = requireNonNull(algorithm, "algorithm");
+        this.validSeconds = validSeconds;
+        this.leewaySeconds = leewaySeconds;
+
+        checkArgument(validSeconds > 0,
+                      "invalid valid duration: " + validSeconds + " (expected: > 0)");
+        checkArgument(leewaySeconds >= 0,
+                      "invalid leeway duration:" + leewaySeconds + " (expected: >= 0)");
+
+        verifier = JWT.require(algorithm)
+                      .withIssuer(issuer)
+                      .acceptLeeway(leewaySeconds)
+                      .build();
+    }
+
+    @Override
+    public String newId() {
+        final Instant now = Instant.now();
+        final int salt = ThreadLocalRandom.current().nextInt(Integer.MAX_VALUE) & 0x7fffffff;
+        return JWT.create()
+                  .withIssuer(issuer)
+                  .withIssuedAt(Date.from(now))
+                  .withExpiresAt(Date.from(now.plus(validSeconds, ChronoUnit.SECONDS)))
+                  .withClaim(CLAIM_NAME_SALT, salt)
+                  .sign(algorithm);
+    }
+
+    @Override
+    public boolean validateId(String id) {
+        requireNonNull(id, "id");
+        try {
+            // Verifier will check whether its issuer is me and also check whether it has been expired or not.
+            final DecodedJWT jwt = verifier.verify(id);
+            return !jwt.getClaim(CLAIM_NAME_SALT).isNull();
```

Comment:
Let's assume that there's no `salt` claim. In this case, we have 3 inputs as follows:
- issuer (string)
- issuedAt (datetime, in seconds)
- expiredAt (datetime, in seconds)

If the 3 inputs are not changed, JWT library produces the same output every time. But, as you see, the 2 time values are in **seconds**, not a **us** or **ms**. So it's possible to produce the same JWT library tokens in a second.

To make JWT library generates different tokens every time, I added the `salt` claim. And I used to random numbers to guess a little harder, but actually it doesn't have no more meaning than making different tokens. So, I just checked only whether the `salt` claim exists here since the token has been validated by `verifier`.

Let me know if you have a concern. :-)

Reply:
> is this a convention in SSO?

There's no convention about the ID in SSO. We need to **generate a new ID** when sending a new authentication request to an IdP(identity provider), and also we need to **validate the ID** when receiving a response from the IdP.

As you might know, SAML request/response is different from the general client-server request/response. They are independent HTTP request(GET or POST). So, in general, we can put the generated ID into the global session storage then validate the received ID by comparing to the generated one. But we want to validate the ID regardless of the global session storage, so I used JWT as an ID of a SAML message.

> Another approach would be to add a claim of id and fill it with UUID and use it rather than the whole token

Yes, it's possible. But I think that it doesn't have an advantage for achieving our goal I explained the above.

> IIUC since the signature is verified, actually there is no extra information from checking the claim is not null.

Yes, you're right. I just added the code for double checking. Let me remove it.

Reply:
Cool makes sense. I like avoiding security checks that don't provide extra information for easier reasoning (I remember one code I've seen that encrypted, hashed, then encrypted again which made it harder to realize that none of the steps used a salt...).

Reply:
To make clear, I renamed the claim from `salt` to `un1`/`un2` which means `uniquifier`. (@trustin 's suggestion)

`un1` is pre-computed when the class is instantiated, using a MAC address and random bytes. It helps to make our `request ID`(a.k.a JWT token) unique in a single machine.

`un2` is the same as the previous `salt` which is generated randomly for every request ID in order to make the ID unique in a second.